### PR TITLE
feat(haystack): add instrumentation support

### PR DESCRIPTION
**Pull request by 21ShisodeParth**
I have read the CLA Document and I hereby sign the CLA

### CODE REVIEW

**Review by axiomofjoy — COMMENTED** [review]
Looks really solid so far. Main feedback:

- no attributes are added in the event of an exception
- Iogic for getting invocation parameters seems incorrect (I think?)

File: python/instrumentation/openinference-instrumentation-haystack/README.md [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+```python
+from openinference.instrumentation.haystack import HaystackInstrumentor
+from opentelemetry import trace as trace_api
+from opentelemetry.exporter.otlp.proto.http.trace_exporter import OTLPSpanExporter
+from opentelemetry.sdk import trace as trace_sdk
+from opentelemetry.sdk.trace.export import SimpleSpanProcessor
+import os
+
+# Set your OpenAI API key
+os.environ["OPENAI_API_KEY"] = "YOUR_KEY_HERE"
```
[the remark this answers is not in the document]
Comment (21ShisodeParth) [reply python/instrumentation/openinference-instrumentation-haystack/README.md:38]: nope! handled thru haystack import

File: python/instrumentation/openinference-instrumentation-haystack/src/openinference/instrumentation/haystack/_wrappers.py [hunk 2]
```diff
@@ -0,0 +1,237 @@
...
+        # Prepare invocation parameters by merging args and kwargs
+        invocation_parameters = {}
+        for arg in args:
+            if arg and isinstance(arg, dict):
+                invocation_parameters.update(arg)
+        invocation_parameters.update(kwargs)
+
+        span_name = component.__class__.__name__
+
+        with self._tracer.start_as_current_span(span_name, attributes={}) as span:
```
[the remark this answers is not in the document]
Comment (21ShisodeParth) [reply python/instrumentation/openinference-instrumentation-haystack/src/openinference/instrumentation/haystack/_wrappers.py:73]: added based on component type; didn't implement via start_as... args but do implement it before checking for an err

File: python/instrumentation/openinference-instrumentation-haystack/src/openinference/instrumentation/haystack/_wrappers.py [hunk 3]
```diff
@@ -0,0 +1,237 @@
...
+
+        if context_api.get_value(context_api._SUPPRESS_INSTRUMENTATION_KEY):
+            return wrapped(*args, **kwargs)
+
+        # Prepare invocation parameters by merging args and kwargs
+        invocation_parameters = {}
+        for arg in args:
+            if arg and isinstance(arg, dict):
+                invocation_parameters.update(arg)
+        invocation_parameters.update(kwargs)
```
[the remark this answers is not in the document]
Comment (21ShisodeParth) [reply python/instrumentation/openinference-instrumentation-haystack/src/openinference/instrumentation/haystack/_wrappers.py:69]: nope! passing in an arg this way updates the invocation parameters dict to include both a key and value (arg name and arg value); i can show you more in depth specific to haystack if you'd like

File: python/instrumentation/openinference-instrumentation-haystack/src/openinference/instrumentation/haystack/_wrappers.py [hunk 4]
```diff
@@ -0,0 +1,254 @@
...
+                _flatten(
+                    {
+                        SpanAttributes.OPENINFERENCE_SPAN_KIND: CHAIN,
+                        SpanAttributes.INPUT_VALUE: safe_json_dumps(invocation_parameters),
+                        SpanAttributes.INPUT_MIME_TYPE: OpenInferenceMimeTypeValues.JSON,
+                    }
+                )
+            ),
+        ) as span:
+            span.set_attributes(dict(get_attributes_from_context()))
```
Comment (axiomofjoy) [inline python/instrumentation/openinference-instrumentation-haystack/src/openinference/instrumentation/haystack/_wrappers.py:237]:
```suggestion
            attributes={
                **_flatten(
                    {
                        SpanAttributes.OPENINFERENCE_SPAN_KIND: CHAIN,
                        SpanAttributes.INPUT_VALUE: safe_json_dumps(invocation_parameters),
                        SpanAttributes.INPUT_MIME_TYPE: OpenInferenceMimeTypeValues.JSON,
                    }
                ),
                **get_attributes_from_context()
            },
        ) as span:
```

Non-blocking suggestion.

File: python/instrumentation/openinference-instrumentation-haystack/src/openinference/instrumentation/haystack/_wrappers.py [hunk 5]
```diff
@@ -0,0 +1,254 @@
...
+            try:
+                response = wrapped(*args, **kwargs)
```
Comment (axiomofjoy) [inline python/instrumentation/openinference-instrumentation-haystack/src/openinference/instrumentation/haystack/_wrappers.py:239]: Are we missing output values here?

File: python/instrumentation/openinference-instrumentation-haystack/src/openinference/instrumentation/haystack/_wrappers.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (axiomofjoy) [reply python/instrumentation/openinference-instrumentation-haystack/src/openinference/instrumentation/haystack/_wrappers.py:72]: Sounds good, I trust you.

File: python/instrumentation/openinference-instrumentation-haystack/src/openinference/instrumentation/haystack/_wrappers.py [hunk 6]
```diff
@@ -0,0 +1,254 @@
...
+        invocation_parameters = {}
+        for arg in args:
+            if arg and isinstance(arg, dict):
+                invocation_parameters.update(arg)
+        invocation_parameters.update(kwargs)
+
+        span_name = component.__class__.__name__
+
+        with self._tracer.start_as_current_span(span_name, attributes={}) as span:
+            if component_type == "llm":
```
Comment (axiomofjoy) [inline python/instrumentation/openinference-instrumentation-haystack/src/openinference/instrumentation/haystack/_wrappers.py:77]: Non-blocking suggestion: I might move this conditional block outside the context manager so you can see the attributes with `self._tracer.start_as_current_span(span_name, attributes=attributes)`

**Review by axiomofjoy — APPROVED** [review]
Nice job! It looks like we are missing output attributes for `Pipeline`. Other than that, LGTM. Might be worth adding a test for the case where this is an error inside a component, but this is non-blocking.

**Pull request merged**